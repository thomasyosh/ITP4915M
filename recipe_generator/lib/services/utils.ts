import { clsx, type ClassValue } from "clsx"
import { twMerge } from "tailwind-merge"

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function getStrapiURL() {
  return process.env.STRAPI_API_URL ?? "https://cms.ite4108m.com";
}