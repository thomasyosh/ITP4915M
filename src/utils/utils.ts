import axios from "axios";
import { getSession } from 'next-auth/react';
import { ExtendedRecipe, PaginationQueryType } from "../types";
import { GetServerSidePropsContext } from "next";

// Filters the results by enhancing recipe information with ownership and liked status for the user
export const filterResults = (recipes: ExtendedRecipe[], userId: string) => {
  return recipes.map((recipe) => (
    {
      ...recipe,
      owner: {
        _id: recipe.owner._id,
        name: recipe.owner.name,
        image: recipe.owner.image
      },
      likedBy: recipe.likedBy.map(({ _id, name, image }) => ({ _id, name, image })), // Simplifies likedBy list
      owns: recipe.owner._id.toString() === userId, // Flags if the recipe belongs to the user
      liked: recipe.likedBy.some(l => l._id.toString() === userId) // Flags if the user liked the recipe
    }
  ))
}

// Updates the recipe list by either replacing or removing a recipe from the list
export const updateRecipeList = (
  oldList: ExtendedRecipe[],
  newRecipe: ExtendedRecipe | null,
  deleteId?: string
) => {
  if (!newRecipe && !deleteId) return oldList
  const id = newRecipe ? newRecipe._id : deleteId;
  return newRecipe
    ? oldList.map(recipe => (recipe._id === id ? newRecipe : recipe))
    : oldList.filter(recipe => recipe._id !== id);
};

// Utility to fetch data on server-side while ensuring user authentication
export const getServerSidePropsUtility = async (context: GetServerSidePropsContext, address: string, propskey: string = 'recipes') => {
  try {
    const session = await getSession(context);
    if (!session) {
      return {
        redirect: {
          destination: '/',
          permanent: false,
        },
      };
    }
    const { data } = await axios.get(`${process.env.NEXT_PUBLIC_API_BASE_URL}/${address}`, {
      headers: {
        Cookie: context.req.headers.cookie || '',
      },
    });
    return {
      props: {
        [propskey]: data,
      },
    };
  } catch (error) {
    console.error(`Failed to fetch ${propskey}:`, error); // Logs errors in fetching data
    return {
      props: {
        [propskey]: [], // Returns an empty list if there's an error
      },
    };
  }
};

// REST API call utility supporting multiple HTTP methods
interface methods {
  put: 'put';
  post: 'post';
  delete: 'delete';
  get: 'get';
}

interface RESTcallTypes {
  address: string;
  method?: keyof methods;
  payload?: {
    [key: string]: any;
  };
}

export const call_api = async ({ address, method = 'get', payload }: RESTcallTypes) => {
  try {
    const { data } = await axios[method as keyof methods](address, payload);
    return data; // Returns the data from the API call
  } catch (error) {
    console.error(`An error occurred making a ${method} REST call to -> ${address} error -> ${error}`);
    throw (error); // Rethrows the error for further handling
  }
};

export const formatDate = (date: string) => {
  const [, day, mth, year] = new Date(date).toUTCString().split(' ');
  return `${day} ${mth} ${year}`;
};

export const playAudio = async (
  audioUrl: string,
  audioRef: React.MutableRefObject<HTMLAudioElement | null>,
  onEnd?: () => void // Optional callback when audio finishes
): Promise<void> => {
  try {
    const audio = new Audio(audioUrl);
    audio.preload = 'auto'; // Force preloading
    audioRef.current = audio; // Save the audio instance

    // Attach the `ended` event listener
    audio.onended = () => {
      if (onEnd) onEnd(); // Call the callback if provided
    };

    // Explicitly force loading
    audio.load();

    // Wait for the audio to preload
    await new Promise<void>((resolve, reject) => {
      let isResolved = false;

      audio.oncanplaythrough = () => {
        if (!isResolved) {
          isResolved = true;
          resolve();
        }
      };

      audio.onerror = () => {
        if (!isResolved) {
          isResolved = true;
          reject(new Error('Error loading audio'));
        }
      };

      setTimeout(() => {
        if (!isResolved) {
          isResolved = true;
          reject(new Error('Audio loading timeout'));
        }
      }, 20000); // 20 seconds timeout
    });

    // Attempt playback
    await audio.play();
  } catch (error: any) {
    console.error(`playAudio: Error playing audio: ${error.message}`);
    throw error;
  }
};

export const paginationQueryHelper = (queryObj: PaginationQueryType) => {
  const page = Number(queryObj.page) || 1;
  const limit = Number(queryObj.limit) || 12; // Default: 12 recipes per page
  const skip = (page - 1) * limit;
  const sortOption = typeof queryObj.sortOption === 'string' ? queryObj.sortOption : 'popular';
  const query = typeof queryObj.query === 'string' ? queryObj.query : undefined;

  return { page, limit, skip, sortOption, query };
};