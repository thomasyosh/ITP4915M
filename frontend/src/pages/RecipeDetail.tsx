import { useRouter } from "next/router";
import Image from 'next/image';
import { HandThumbUpIcon, CheckCircleIcon } from '@heroicons/react/24/outline';
import { EllipsisHorizontalIcon, HandThumbUpIcon as HandThumbUpSolid } from '@heroicons/react/16/solid'
import useActionPopover from "../components/Hooks/useActionPopover";
import { useRecipeData } from "../components/Hooks/useRecipeData";
import { ActionPopover } from "../components/Recipe_Display/ActionPopover";
import RecipeHeader from "../components/RecipeHeader";
import UserLink from "../components/UserLink";
import Loading from "../components/Loading";
import ErrorPage from "./auth/error";
import { call_api } from "../utils/utils";

const getThumbsup = ({ liked, owns }: { liked: boolean, owns: boolean }) => {
    const baseClass = "size-8";
    if (owns) {
        return <HandThumbUpSolid className={`${baseClass} text-gray-500`} />;
    }
    if (liked) {
        return <HandThumbUpSolid className={`${baseClass} text-blue-500`} />;
    }
    return <HandThumbUpIcon className={`${baseClass} text-blue-500`} />;
};

export default function RecipeDetail() {
    const router = useRouter();
    const { recipeId } = router.query as { recipeId?: string }; // Extract recipeId from the URL query parameters
    const { recipeData, loading, error, setRecipeData, setLoading } = useRecipeData(recipeId);


    const updateRecipe = (audioLink: string) => {
        setRecipeData((prevRecipdata) => {
            if (!prevRecipdata) return null; // Ensure we're not working with null

            return {
                ...prevRecipdata, // Keep all existing properties
                audio: audioLink, // Update only what's necessary
            };
        });
    }

    const handleRecipeLike = async (recipeId: string) => {
        try {
            const result = await call_api({ address: '/api/like-recipe', method: 'put', payload: { recipeId } })
            setRecipeData(result)
        } catch (error) {
            console.log(error)
        }
    }

    const {
        handleClone,
        handleCopy,
        handlePlayRecipe,
        handleDeleteDialog,
        handleDeleteRecipe,
        linkCopied,
        isPlayingAudio,
        isLoadingAudio,
        isDeleteDialogOpen
    } = useActionPopover(recipeData, updateRecipe);



    const deleteAndRemoveRecipe = async () => {
        try {
            setLoading(true)
            const { message, error } = await handleDeleteRecipe();
            setLoading(false)
            router.push('/')
        } catch (error) {
            console.error(error)
        }
    }

    // Render the ErrorPage component if recipeId is not present
    if (!recipeId) return <ErrorPage message="Invalid Recipe" />;
    // Render the Loading component while data is being fetched
    if (loading) return <Loading />;
    // Render the ErrorPage component if an error occurred during fetching
    if (error) return <ErrorPage message={error} />;
    // Render a fallback message if no recipe data is found
    if (!recipeData) return <ErrorPage message="No Recipe Data" />;

    return (
        <div className="min-h-screen bg-gray-100 p-4">
            {/* Recipe Card */}
            <div className="max-w-4xl mx-auto bg-white shadow-lg rounded-lg overflow-hidden">
                <RecipeHeader recipeData={recipeData} /> {/* Recipe header with image and title */}
                <div className="p-6">
                    <ActionPopover
                        handlers={{
                            handleClone,
                            handleCopy,
                            deleteDialog: handleDeleteDialog,
                            handlePlayRecipe,
                            deleteRecipe: deleteAndRemoveRecipe,
                        }}
                        states={{
                            hasAudio: Boolean(recipeData.audio),
                            isLoadingAudio,
                            isPlayingAudio,
                            linkCopied,
                            isDeleteDialogOpen,
                        }}
                        data={{
                            recipe: recipeData,
                            buttonType: <EllipsisHorizontalIcon className="h-6 w-6 text-gray-700" />
                        }}

                    />
                    {/* Ingredients */}
                    <div className="flex flex-col">
                        <div className="mb-4">
                            <h3 className="text-xl font-semibold mb-2">Ingredients</h3> {/* Section title */}
                            <ul className="grid grid-cols-1 sm:grid-cols-2 gap-2"> {/* Responsive grid layout */}
                                {recipeData.ingredients.map((ingredient) => (
                                    <li key={ingredient.name} className="flex items-center"> {/* Ingredient item */}
                                        <CheckCircleIcon className="w-5 h-5 text-green-500 mr-2 flex-shrink-0" /> {/* Icon next to ingredient */}
                                        <span className="text-gray-700">
                                            {ingredient.name}{ingredient.quantity && ` (${ingredient.quantity})`} {/* Ingredient name and quantity */}
                                        </span>
                                    </li>
                                ))}
                            </ul>
                        </div>
                    </div>


                    {/* Instructions */}
                    <div className="mb-4">
                        <h3 className="text-xl font-semibold mb-2">Instructions</h3> {/* Section title */}
                        <ol className="list-decimal list-inside space-y-4"> {/* Ordered list with spacing */}
                            {recipeData.instructions.map((step, index) => (
                                <li key={index} className="flex items-start"> {/* Instruction step */}
                                    <span className="font-bold text-gray-800 mr-3">{index + 1}.</span> {/* Step number */}
                                    <p className="text-gray-700">{step}</p> {/* Instruction text */}
                                </li>
                            ))}
                        </ol>
                    </div>

                    {/* Additional Information */}
                    <div className="mb-6">
                        <h3 className="text-xl font-semibold mb-4 text-gray-800">Additional Information</h3> {/* Section title */}
                        <div className="space-y-4"> {/* Spacing between info items */}
                            <div>
                                <h4 className="font-bold text-gray-700">Tips:</h4> {/* Subsection title */}
                                <p className="text-gray-600">{recipeData.additionalInformation.tips}</p> {/* Tips content */}
                            </div>
                            <div>
                                <h4 className="font-bold text-gray-700">Variations:</h4> {/* Subsection title */}
                                <p className="text-gray-600">{recipeData.additionalInformation.variations}</p> {/* Variations content */}
                            </div>
                            <div>
                                <h4 className="font-bold text-gray-700">Serving Suggestions:</h4> {/* Subsection title */}
                                <p className="text-gray-600">{recipeData.additionalInformation.servingSuggestions}</p> {/* Serving suggestions content */}
                            </div>
                            <div>
                                <h4 className="font-bold text-gray-700">Nutritional Information:</h4> {/* Subsection title */}
                                <p className="text-gray-600">{recipeData.additionalInformation.nutritionalInformation}</p> {/* Nutritional information content */}
                            </div>
                        </div>
                    </div>

                    {/* Liked By Section */}
                    <button
                        className="w-14 h-14 mb-3 hover:text-blue-600 hover:scale-105 hover:shadow rounded-full flex items-center justify-center"
                        onClick={() => handleRecipeLike(recipeData._id)}
                        disabled={recipeData.owns}
                        data-testid="like_button"
                    >
                        {getThumbsup(recipeData)}
                    </button>
                    {/* Thumbs up icon */}
                    <div className="flex items-center">
                        <div className="flex flex-wrap items-center gap-2">
                            {recipeData.likedBy.map((user, idx) => (
                                <div key={user._id} className="flex items-center space-x-2"> {/* User item */}
                                    <div className="relative w-8 h-8 rounded-full overflow-hidden"> {/* User avatar container */}
                                        <Image
                                            src={user.image} // User's image source
                                            alt={user.name} // Alt text for accessibility
                                            fill // Fill the parent container
                                            style={{ objectFit: 'cover' }} // Ensure the image covers the container
                                            className="rounded-full" // Make the image circular
                                        />
                                    </div>
                                    <span className="text-gray-700"><UserLink name={user.name} userId={user._id}/></span>
                                </div>
                            ))}
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
}

