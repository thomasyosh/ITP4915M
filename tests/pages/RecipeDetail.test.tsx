import RecipeDetail from "../../src/pages/RecipeDetail";
import { render, screen, fireEvent } from '@testing-library/react'
import { useRouter } from 'next/router';
import * as apiCalls from "../../src/utils/utils";
import { stub_recipe_1 } from '../stub'

const routePushMock = jest.fn()

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({
        pathName: 'mocked Path',
        push: routePushMock,
        query: {
            recipeId: 'stub_recipeid'
        },
        events:{
            on: jest.fn(),
            off: jest.fn()
        }
    })),
}))

jest.mock("next/image")

jest.mock("../../src/utils/utils", () => ({
    ...jest.requireActual("../../src/utils/utils"),
    call_api: jest.fn()
}))

describe('The Recipe Detail Page', () => {
    it('shall succesfully render a recipe', async () => {
        //mock api call
        const getRecipeFromAPI = jest.spyOn(apiCalls, 'call_api');
        getRecipeFromAPI.mockImplementationOnce(() => Promise.resolve(stub_recipe_1))
        const { container } = render(
            <RecipeDetail />
        )
        expect(await screen.findByText('Recipe_1_name')).toBeInTheDocument()
        expect(container).toMatchSnapshot()
    })
    it('shall render the error page if no recipe id', async () => {
        //mock query route
        (useRouter as jest.Mock).mockImplementationOnce(() => ({
            query: {}
        }))
        const { container } = render(
            <RecipeDetail />
        )
        expect(await screen.findByText('Invalid Recipe')).toBeInTheDocument()
    })
    it('shall render the error page if there is a problem with fetching data', async () => {
         //mock api call
        const getRecipeFromAPI = jest.spyOn(apiCalls, 'call_api');
        getRecipeFromAPI.mockImplementationOnce(() => Promise.reject({message: 'Fetch error message'}))
        const { container } = render(
            <RecipeDetail />
        )
        expect(await screen.findByText('Fetch error message')).toBeInTheDocument()
    })

    it('shall render the error page if no data is returned', async () => {
         //mock api call
        const getRecipeFromAPI = jest.spyOn(apiCalls, 'call_api');
        getRecipeFromAPI.mockImplementationOnce(() => Promise.resolve(undefined))
        const { container } = render(
            <RecipeDetail />
        )
        expect(await screen.findByText('No Recipe Data')).toBeInTheDocument()
    })
})
