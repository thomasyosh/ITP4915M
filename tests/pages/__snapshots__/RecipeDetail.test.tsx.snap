// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Recipe Detail Page shall succesfully render a recipe 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-100 p-4"
  >
    <div
      class="max-w-4xl mx-auto bg-white shadow-lg rounded-lg overflow-hidden"
    >
      <div
        class="relative w-full h-80"
      />
      <div
        class="pl-6 pt-6 pr-6"
      >
        <h2
          class="text-2xl font-bold mb-2 mt-2"
        >
          Recipe_1_name
        </h2>
         
        <div
          class="flex items-center mb-6"
        >
          <div
            class="relative w-12 h-12 rounded-full overflow-hidden mr-4"
          />
          <div>
            <span
              class="text-gray-700 text-lg"
            >
              By 
              <span
                class="text-gray-600 hover:text-gray-800 hover:underline cursor-pointer"
              >
                user_1
              </span>
            </span>
            <p
              class="text-sm text-gray-500"
            >
              02 Jul 2024
            </p>
          </div>
        </div>
        <div
          class="mb-0"
        >
          <h3
            class="text-xl font-semibold mb-2"
          >
            Dietary Preferences
          </h3>
           
          <div
            class="flex flex-wrap gap-2"
          >
             
            <span
              class="bg-purple-100 text-purple-800 text-sm font-medium px-3 py-1 rounded-full"
            >
              Recipe_1_preference_1
               
            </span>
            <span
              class="bg-purple-100 text-purple-800 text-sm font-medium px-3 py-1 rounded-full"
            >
              Recipe_1_preference_2
               
            </span>
          </div>
        </div>
      </div>
       
      <div
        class="p-6"
      >
        <div
          class="relative"
          data-headlessui-state=""
        >
          <button
            aria-expanded="false"
            class="flex items-center justify-center w-12 h-12 ml-auto bg-gray-100 rounded-full hover:bg-gray-200 focus:outline-none focus:ring-2 focus:ring-gray-300 active:bg-gray-300 transition-all duration-200"
            data-headlessui-state=""
            id="headlessui-popover-button-:r2:"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="h-6 w-6 text-gray-700"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2 8a1.5 1.5 0 1 1 3 0 1.5 1.5 0 0 1-3 0ZM6.5 8a1.5 1.5 0 1 1 3 0 1.5 1.5 0 0 1-3 0ZM12.5 6.5a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3Z"
              />
            </svg>
          </button>
        </div>
        <div
          class="flex flex-col"
        >
          <div
            class="mb-4"
          >
            <h3
              class="text-xl font-semibold mb-2"
            >
              Ingredients
            </h3>
             
            <ul
              class="grid grid-cols-1 sm:grid-cols-2 gap-2"
            >
               
              <li
                class="flex items-center"
              >
                 
                <svg
                  aria-hidden="true"
                  class="w-5 h-5 text-green-500 mr-2 flex-shrink-0"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 12.75 11.25 15 15 9.75M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                 
                <span
                  class="text-gray-700"
                >
                  Recipe_1_Ingredient_1
                   (Recipe_1_Ingredient_1_quantity_1)
                   
                </span>
              </li>
              <li
                class="flex items-center"
              >
                 
                <svg
                  aria-hidden="true"
                  class="w-5 h-5 text-green-500 mr-2 flex-shrink-0"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 12.75 11.25 15 15 9.75M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                 
                <span
                  class="text-gray-700"
                >
                  Recipe_1_Ingredient_2
                   (Recipe_1_Ingredient_2_quantity_2)
                   
                </span>
              </li>
            </ul>
          </div>
        </div>
        <div
          class="mb-4"
        >
          <h3
            class="text-xl font-semibold mb-2"
          >
            Instructions
          </h3>
           
          <ol
            class="list-decimal list-inside space-y-4"
          >
             
            <li
              class="flex items-start"
            >
               
              <span
                class="font-bold text-gray-800 mr-3"
              >
                1
                .
              </span>
               
              <p
                class="text-gray-700"
              >
                Recipe_1_Instructions_1.
              </p>
               
            </li>
            <li
              class="flex items-start"
            >
               
              <span
                class="font-bold text-gray-800 mr-3"
              >
                2
                .
              </span>
               
              <p
                class="text-gray-700"
              >
                Recipe_1_Instructions_2.
              </p>
               
            </li>
          </ol>
        </div>
        <div
          class="mb-6"
        >
          <h3
            class="text-xl font-semibold mb-4 text-gray-800"
          >
            Additional Information
          </h3>
           
          <div
            class="space-y-4"
          >
             
            <div>
              <h4
                class="font-bold text-gray-700"
              >
                Tips:
              </h4>
               
              <p
                class="text-gray-600"
              >
                Recipe_1_tips
              </p>
               
            </div>
            <div>
              <h4
                class="font-bold text-gray-700"
              >
                Variations:
              </h4>
               
              <p
                class="text-gray-600"
              >
                Recipe_1_variations
              </p>
               
            </div>
            <div>
              <h4
                class="font-bold text-gray-700"
              >
                Serving Suggestions:
              </h4>
               
              <p
                class="text-gray-600"
              >
                Recipe_1_servingSuggestions
              </p>
               
            </div>
            <div>
              <h4
                class="font-bold text-gray-700"
              >
                Nutritional Information:
              </h4>
               
              <p
                class="text-gray-600"
              >
                Recipe_1_nutritionalInformation
              </p>
               
            </div>
          </div>
        </div>
        <button
          class="w-14 h-14 mb-3 hover:text-blue-600 hover:scale-105 hover:shadow rounded-full flex items-center justify-center"
          data-testid="like_button"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="size-8 text-gray-500"
            data-slot="icon"
            fill="currentColor"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.09 15a1 1 0 0 0 1-1V8a1 1 0 1 0-2 0v6a1 1 0 0 0 1 1ZM5.765 13H4.09V8c.663 0 1.218-.466 1.556-1.037a4.02 4.02 0 0 1 1.358-1.377c.478-.292.907-.706.989-1.26V4.32a9.03 9.03 0 0 0 0-2.642c-.028-.194.048-.394.224-.479A2 2 0 0 1 11.09 3c0 .812-.08 1.605-.235 2.371a.521.521 0 0 0 .502.629h1.733c1.104 0 2.01.898 1.901 1.997a19.831 19.831 0 0 1-1.081 4.788c-.27.747-.998 1.215-1.793 1.215H9.414c-.215 0-.428-.035-.632-.103l-2.384-.794A2.002 2.002 0 0 0 5.765 13Z"
            />
          </svg>
        </button>
        <div
          class="flex items-center"
        >
          <div
            class="flex flex-wrap items-center gap-2"
          >
            <div
              class="flex items-center space-x-2"
            >
               
              <div
                class="relative w-8 h-8 rounded-full overflow-hidden"
              >
                 
              </div>
              <span
                class="text-gray-700"
              >
                <span
                  class="text-gray-600 hover:text-gray-800 hover:underline cursor-pointer"
                >
                  user_2
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
